<!--
  ~ Copyright (c) 2013, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->

<!--
  ~ Copyright (c) 2012, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->



<?php $group = 'Queue Scripts'; ?>
        <?php ob_start(); ?>
        <!DOCTYPE html>
        <html lang='en'>
        <head>
            <meta charset='utf-8'>
            <title>SVQScript documentation documentation</title>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <meta name='description' content=''>

            <link type="text/css" rel="stylesheet" href="bootstrap.min.css">
            <script src="jquery.min.js"></script>
            <script src="bootstrap.min.js"></script>
            <style>
                @media (min-width: 980px) {
                    body {
                        padding-top : 60px;
                    }
                }
            </style>

        </head>
        <body>
        <div class="navbar navbar-fixed-top">
            <div class="navbar-inner">
                <div class="container-fluid">
                    <a class="brand" href="index.html">GenomeSTRiP Documentation</a>
                </div>
            </div>
        </div>
        <div class="container-fluid">
            <?php
    // Output Buffering allows for the website to inject a different header
    ob_clean();
    include '../include/common.php';
    printHeader("SVQScript documentation", "Guides");
    ?>

            <div class="row-fluid">

<section class="span4">
    <aside class="well">
        <h2>Categories</h2>
        <style>
            #sidenav .accordion-body a {
                color : gray;
            }

            .accordion-body li {
                list-style : none;
            }
        </style>
        <ul class="nav nav-pills nav-stacked" id="sidenav">
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#QueueScripts">Queue Scripts</a>

                    <div id="QueueScripts"
                    <?php echo ($group == 'Queue Scripts')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVDiscovery.html">SVDiscovery</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVGenotyper.html">SVGenotyper</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVPreprocess.html">SVPreprocess</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVQScript.html">SVQScript</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotation">Variant Annotation</a>

                    <div id="VariantAnnotation"
                    <?php echo ($group == 'Variant Annotation')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_SVVariantAnnotatorWalker.html">SVVariantAnnotatorWalker</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotators">Variant Annotators</a>

                    <div id="VariantAnnotators"
                    <?php echo ($group == 'Variant Annotators')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_AlleleFrequencyAnnotator.html">AlleleFrequencyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ClusterSeparationAnnotator.html">ClusterSeparationAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_CopyNumberClassAnnotator.html">CopyNumberClassAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GCContentAnnotator.html">GCContentAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GeneOverlapAnnotator.html">GeneOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_IntensityRankSumAnnotator.html">IntensityRankSumAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_NonVariantAnnotator.html">NonVariantAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ReciprocalOverlapAnnotator.html">ReciprocalOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_RedundancyAnnotator.html">RedundancyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_TagVariantsAnnotator.html">TagVariantsAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_VariantsPerSampleAnnotator.html">VariantsPerSampleAnnotator</a>
                                </li>
                    </ul>
                    </div>
                </li>
        </ul>
    </aside>
    <?php getForumPosts( 'SVQScript' ) ?>

</section>

<div class="span8">

    <h1>SVQScript</h1>

    <p class="lead">Queue utilities for writing structural variation discovery/genotyping pipepines.</p>
        <h3>Category
            <small> Queue Scripts</small>
        </h3>
    <br/>

    <h2>Introduction</h2>
    

            <hr>
            <h3>SVQScript specific arguments</h3>
            <table class="table table-striped table-bordered table-condensed">
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Type</th>
                    <th>Default value</th>
                    <th>Summary</th>
                </tr>
                </thead>
                <tbody>
        <tr>
            <th colspan="4" id="row-divider">Required</th>
        </tr>
            <tr>
                <td><a href="#--gatkjar">--gatkjar</a></td>
                <td>File</td>
                <td>NA</td>
                <td>The path to the GenomeAnalysisTK.jar file</td>
            </tr>
            <tr>
                <td><a href="#--referencefile">--referencefile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>The reference file for the bam files</td>
            </tr>
        <tr>
            <th colspan="4" id="row-divider">Optional</th>
        </tr>
            <tr>
                <td><a href="#--alignunmappedmates">--alignunmappedmates</a></td>
                <td>Boolean</td>
                <td>NA</td>
                <td>Whether to align unmapped mates to alternate alleles (default true)</td>
            </tr>
            <tr>
                <td><a href="#--altAlleleAlignments">--altAlleleAlignments</a></td>
                <td>List[File]</td>
                <td>NA</td>
                <td>Alternate allele alignments</td>
            </tr>
            <tr>
                <td><a href="#--altalleleflanklength">--altalleleflanklength</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Length of flanking sequence to use for alt allele mapping</td>
            </tr>
            <tr>
                <td><a href="#--altmaskfiles">--altmaskfiles</a></td>
                <td>List[File]</td>
                <td>NA</td>
                <td>The genome mask file(s) for the alternate alleles</td>
            </tr>
            <tr>
                <td><a href="#--bamfilesaredisjoint">--bamfilesaredisjoint</a></td>
                <td>Boolean</td>
                <td>NA</td>
                <td>Reduces memory footprint if each input BAM file contains separate samples/libraries/read groups</td>
            </tr>
            <tr>
                <td><a href="#--baminputs">--baminputs</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>One or more bam files or a bam file list (with extension .list)</td>
            </tr>
            <tr>
                <td><a href="#--batchsystem">--batchsystem</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Batch system to be used when submitting scripts from within this script (bsub/qsub)</td>
            </tr>
            <tr>
                <td><a href="#--classpath">--classpath</a></td>
                <td>String</td>
                <td>NA</td>
                <td>The java classpath</td>
            </tr>
            <tr>
                <td><a href="#--configfile">--configfile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Configuration property file</td>
            </tr>
            <tr>
                <td><a href="#--copynumbermaskfile">--copynumbermaskfile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Genome mask file for regions of potentially polymorphic copy number</td>
            </tr>
            <tr>
                <td><a href="#--debug">--debug</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Enable verbose output for debugging</td>
            </tr>
            <tr>
                <td><a href="#--depthmaximuminsertsizeradius">--depthmaximuminsertsizeradius</a></td>
                <td>Double</td>
                <td>NA</td>
                <td>Maximum insert size radius for read depth analysis</td>
            </tr>
            <tr>
                <td><a href="#--depthminimummappingquality">--depthminimummappingquality</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Minimum mapping quality for read depth analysis</td>
            </tr>
            <tr>
                <td><a href="#--disablegatktraversal">--disablegatktraversal</a></td>
                <td>Boolean</td>
                <td>NA</td>
                <td>Disable GATK normal walker traversal</td>
            </tr>
            <tr>
                <td><a href="#--discoveryInterval">--discoveryInterval</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Interval(s) for SV discovery (or file of intervals)</td>
            </tr>
            <tr>
                <td><a href="#--discoveryMaximumSize">--discoveryMaximumSize</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Maximum event size for SV discovery</td>
            </tr>
            <tr>
                <td><a href="#--discoveryMinimumSize">--discoveryMinimumSize</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Minimum event size for SV discovery</td>
            </tr>
            <tr>
                <td><a href="#--discoveryWindowPadding">--discoveryWindowPadding</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Window padding for SV discovery</td>
            </tr>
            <tr>
                <td><a href="#--discoveryWindowSize">--discoveryWindowSize</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Window size for SV discovery</td>
            </tr>
            <tr>
                <td><a href="#--duplicateoverlapthreshold">--duplicateoverlapthreshold</a></td>
                <td>Double</td>
                <td>NA</td>
                <td>Overlap threshold for duplicate site filtering (default 0.5)</td>
            </tr>
            <tr>
                <td><a href="#--duplicatescorethreshold">--duplicatescorethreshold</a></td>
                <td>Double</td>
                <td>NA</td>
                <td>Concordance LOD score threshold (minimum per-sample LOD score at which events are considered non-redundant, default zero)</td>
            </tr>
            <tr>
                <td><a href="#--excludereadgroup">--excludereadgroup</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Read group(s) to exclude or .list files of read groups</td>
            </tr>
            <tr>
                <td><a href="#--exomedepthmatrixfile">--exomedepthmatrixfile</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Input data matrix for exome genotyping (experimental)</td>
            </tr>
            <tr>
                <td><a href="#--exomemode">--exomemode</a></td>
                <td>String</td>
                <td>NA</td>
                <td>True to enable exome genotyping mode (experimental) (default is false)</td>
            </tr>
            <tr>
                <td><a href="#--gendermapfile">--gendermapfile</a></td>
                <td>List[File]</td>
                <td>NA</td>
                <td>File(s) containing the declared gender for each sample</td>
            </tr>
            <tr>
                <td><a href="#--genomemaskfiles">--genomemaskfiles</a></td>
                <td>List[File]</td>
                <td>NA</td>
                <td>The genome mask file(s) for the reference file</td>
            </tr>
            <tr>
                <td><a href="#--inputfileindexcache">--inputfileindexcache</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Location where the input BAM files' indices are cached</td>
            </tr>
            <tr>
                <td><a href="#--jobLogDirectory">--jobLogDirectory</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Directory for queue output files</td>
            </tr>
            <tr>
                <td><a href="#--lsfdefaultmemlimit">--lsfdefaultmemlimit</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Override default LSF MEMLIMIT setting (in GB)</td>
            </tr>
            <tr>
                <td><a href="#--lsfresource">--lsfresource</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Resource request rusage string to supply to LSF</td>
            </tr>
            <tr>
                <td><a href="#--metaDataLocation">--metaDataLocation</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Path to meta-data directory</td>
            </tr>
            <tr>
                <td><a href="#--minimuminsertsizemapfile">--minimuminsertsizemapfile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>File containing a manual overrides of minimum insert size used to select read pairs</td>
            </tr>
            <tr>
                <td><a href="#--nonvariantscorethreshold">--nonvariantscorethreshold</a></td>
                <td>Double</td>
                <td>NA</td>
                <td>Score threshold for non-variant site filtering (phred scale, default 13.0)</td>
            </tr>
            <tr>
                <td><a href="#--paralleljobs">--paralleljobs</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Number of parallel jobs to run</td>
            </tr>
            <tr>
                <td><a href="#--parallelrecords">--parallelrecords</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Number of records to process in each parallel job</td>
            </tr>
            <tr>
                <td><a href="#--parameter">--parameter</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Override individual parameters from the configuration property file</td>
            </tr>
            <tr>
                <td><a href="#--ploidymapfile">--ploidymapfile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Map file defining the gender-specific ploidy of each region of the reference sequence</td>
            </tr>
            <tr>
                <td><a href="#--rundirectory">--rundirectory</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Directory for auxilliary output files</td>
            </tr>
            <tr>
                <td><a href="#--samplelist">--samplelist</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Sample(s) to process or .list files of samples</td>
            </tr>
            <tr>
                <td><a href="#--suppressvcfcommandlines">--suppressvcfcommandlines</a></td>
                <td>Boolean</td>
                <td>NA</td>
                <td>Flag indicating whether to suppress the command line header in the output VCF file</td>
            </tr>
            <tr>
                <td><a href="#--varianttype">--varianttype</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Variant discovery</td>
            </tr>
            <tr>
                <td><a href="#--vcfcachingdistance">--vcfcachingdistance</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Caching distance for VCF records to preserve sort order (default 10K)</td>
            </tr>
                </tbody>
            </table>



                        <hr>
                            <h2>Argument details</h2>
    <h3><a name="--alignunmappedmates">--alignunmappedmates
         / -alignUnmappedMates
    </a>
        (
        
        Boolean
        
        )
    </h3>
    <p class="args">
        Whether to align unmapped mates to alternate alleles (default true). 
            </p>
    <h3><a name="--altAlleleAlignments">--altAlleleAlignments
         / -altAlignments
    </a>
        (
        
        List[File]
        
        )
    </h3>
    <p class="args">
        Alternate allele alignments. 
            </p>
    <h3><a name="--altalleleflanklength">--altalleleflanklength
         / -altAlleleFlankLength
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Length of flanking sequence to use for alt allele mapping. 
            </p>
    <h3><a name="--altmaskfiles">--altmaskfiles
         / -altMaskFile
    </a>
        (
        
        List[File]
        
        )
    </h3>
    <p class="args">
        The genome mask file(s) for the alternate alleles. 
            </p>
    <h3><a name="--bamfilesaredisjoint">--bamfilesaredisjoint
         / -bamFilesAreDisjoint
    </a>
        (
        
        Boolean
        
        )
    </h3>
    <p class="args">
        Reduces memory footprint if each input BAM file contains separate samples/libraries/read groups. 
            </p>
    <h3><a name="--baminputs">--baminputs
         / -I
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        One or more bam files or a bam file list (with extension .list). 
            </p>
    <h3><a name="--batchsystem">--batchsystem
         / -batchSystem
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Batch system to be used when submitting scripts from within this script (bsub/qsub). 
            </p>
    <h3><a name="--classpath">--classpath
         / -cp
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        The java classpath. 
            </p>
    <h3><a name="--configfile">--configfile
         / -configFile
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Configuration property file. 
            </p>
    <h3><a name="--copynumbermaskfile">--copynumbermaskfile
         / -copyNumberMaskFile
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Genome mask file for regions of potentially polymorphic copy number. 
            </p>
    <h3><a name="--debug">--debug
         / -debug
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Enable verbose output for debugging. 
            </p>
    <h3><a name="--depthmaximuminsertsizeradius">--depthmaximuminsertsizeradius
         / -depthMaximumInsertSizeRadius
    </a>
        (
        
        Double
        
        )
    </h3>
    <p class="args">
        Maximum insert size radius for read depth analysis. 
            </p>
    <h3><a name="--depthminimummappingquality">--depthminimummappingquality
         / -depthMinimumMappingQuality
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Minimum mapping quality for read depth analysis. 
            </p>
    <h3><a name="--disablegatktraversal">--disablegatktraversal
         / -disableGATKTraversal
    </a>
        (
        
        Boolean
        
        )
    </h3>
    <p class="args">
        Disable GATK normal walker traversal. 
            </p>
    <h3><a name="--discoveryInterval">--discoveryInterval
         / -L
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Interval(s) for SV discovery (or file of intervals). 
            </p>
    <h3><a name="--discoveryMaximumSize">--discoveryMaximumSize
         / -maximumSize
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Maximum event size for SV discovery. 
            </p>
    <h3><a name="--discoveryMinimumSize">--discoveryMinimumSize
         / -minimumSize
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Minimum event size for SV discovery. 
            </p>
    <h3><a name="--discoveryWindowPadding">--discoveryWindowPadding
         / -windowPadding
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Window padding for SV discovery. 
            </p>
    <h3><a name="--discoveryWindowSize">--discoveryWindowSize
         / -windowSize
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Window size for SV discovery. 
            </p>
    <h3><a name="--duplicateoverlapthreshold">--duplicateoverlapthreshold
         / -duplicateOverlapThreshold
    </a>
        (
        
        Double
        
        )
    </h3>
    <p class="args">
        Overlap threshold for duplicate site filtering (default 0.5). 
            </p>
    <h3><a name="--duplicatescorethreshold">--duplicatescorethreshold
         / -duplicateScoreThreshold
    </a>
        (
        
        Double
        
        )
    </h3>
    <p class="args">
        Concordance LOD score threshold (minimum per-sample LOD score at which events are considered non-redundant, default zero). 
            </p>
    <h3><a name="--excludereadgroup">--excludereadgroup
         / -excludeReadGroup
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Read group(s) to exclude or .list files of read groups. 
            </p>
    <h3><a name="--exomedepthmatrixfile">--exomedepthmatrixfile
         / -exomeDepthMatrixFile
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Input data matrix for exome genotyping (experimental). 
            </p>
    <h3><a name="--exomemode">--exomemode
         / -exome
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        True to enable exome genotyping mode (experimental) (default is false). 
            </p>
    <h3><a name="--gatkjar">--gatkjar
         / -gatk
    </a>
        (
        required
        File
        
        )
    </h3>
    <p class="args">
        The path to the GenomeAnalysisTK.jar file. 
            </p>
    <h3><a name="--gendermapfile">--gendermapfile
         / -genderMapFile
    </a>
        (
        
        List[File]
        
        )
    </h3>
    <p class="args">
        File(s) containing the declared gender for each sample. 
            </p>
    <h3><a name="--genomemaskfiles">--genomemaskfiles
         / -genomeMaskFile
    </a>
        (
        
        List[File]
        
        )
    </h3>
    <p class="args">
        The genome mask file(s) for the reference file. 
            </p>
    <h3><a name="--inputfileindexcache">--inputfileindexcache
         / -inputFileIndexCache
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Location where the input BAM files' indices are cached. 
            </p>
    <h3><a name="--jobLogDirectory">--jobLogDirectory
         / -jobLogDir
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Directory for queue output files. 
            </p>
    <h3><a name="--lsfdefaultmemlimit">--lsfdefaultmemlimit
         / -lsfMemLimit
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Override default LSF MEMLIMIT setting (in GB). 
            </p>
    <h3><a name="--lsfresource">--lsfresource
         / -lsfResource
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Resource request rusage string to supply to LSF. 
            </p>
    <h3><a name="--metaDataLocation">--metaDataLocation
         / -md
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Path to meta-data directory. 
            </p>
    <h3><a name="--minimuminsertsizemapfile">--minimuminsertsizemapfile
         / -minimumInsertSizeMapFile
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        File containing a manual overrides of minimum insert size used to select read pairs. 
            </p>
    <h3><a name="--nonvariantscorethreshold">--nonvariantscorethreshold
         / -nonVariantScoreThreshold
    </a>
        (
        
        Double
        
        )
    </h3>
    <p class="args">
        Score threshold for non-variant site filtering (phred scale, default 13.0). 
            </p>
    <h3><a name="--paralleljobs">--paralleljobs
         / -parallelJobs
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Number of parallel jobs to run. 
            </p>
    <h3><a name="--parallelrecords">--parallelrecords
         / -parallelRecords
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Number of records to process in each parallel job. 
            </p>
    <h3><a name="--parameter">--parameter
         / -P
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Override individual parameters from the configuration property file. 
            </p>
    <h3><a name="--ploidymapfile">--ploidymapfile
         / -ploidyMapFile
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Map file defining the gender-specific ploidy of each region of the reference sequence. 
            </p>
    <h3><a name="--referencefile">--referencefile
         / -R
    </a>
        (
        required
        File
        
        )
    </h3>
    <p class="args">
        The reference file for the bam files. 
            </p>
    <h3><a name="--rundirectory">--rundirectory
         / -runDirectory
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Directory for auxilliary output files. 
            </p>
    <h3><a name="--samplelist">--samplelist
         / -sample
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Sample(s) to process or .list files of samples. 
            </p>
    <h3><a name="--suppressvcfcommandlines">--suppressvcfcommandlines
         / -suppressVCFCommandLines
    </a>
        (
        
        Boolean
        
        )
    </h3>
    <p class="args">
        Flag indicating whether to suppress the command line header in the output VCF file. 
            </p>
    <h3><a name="--varianttype">--varianttype
         / -variantType
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Variant discovery. 
            </p>
    <h3><a name="--vcfcachingdistance">--vcfcachingdistance
         / -vcfCachingDistance
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Caching distance for VCF records to preserve sort order (default 10K). 
            </p>

        <hr>
        <p class="see-also">See also <a href="index.html">Documentation index</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip">GenomeSTRiP Site</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip/documentation">GenomeSTRiP Support</a></p>

        <p class="version">GenomeSTRiP version 1.04-r1441 built at 2014/07/14 14:26:06.</p>
        </div><!--/.span-->
        </div><!--/.row-->
        </body>
        <?php printFooter();
        ob_end_flush();
        ?>
