<!--
  ~ Copyright (c) 2013, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->

<!--
  ~ Copyright (c) 2012, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->



<?php $group = 'Variant Annotators'; ?>
        <?php ob_start(); ?>
        <!DOCTYPE html>
        <html lang='en'>
        <head>
            <meta charset='utf-8'>
            <title>IntensityRankSumAnnotator documentation documentation</title>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <meta name='description' content=''>

            <link type="text/css" rel="stylesheet" href="bootstrap.min.css">
            <script src="jquery.min.js"></script>
            <script src="bootstrap.min.js"></script>
            <style>
                @media (min-width: 980px) {
                    body {
                        padding-top : 60px;
                    }
                }
            </style>

        </head>
        <body>
        <div class="navbar navbar-fixed-top">
            <div class="navbar-inner">
                <div class="container-fluid">
                    <a class="brand" href="index.html">GenomeSTRiP Documentation</a>
                </div>
            </div>
        </div>
        <div class="container-fluid">
            <?php
    // Output Buffering allows for the website to inject a different header
    ob_clean();
    include '../include/common.php';
    printHeader("IntensityRankSumAnnotator documentation", "Guides");
    ?>

            <div class="row-fluid">

<section class="span4">
    <aside class="well">
        <h2>Categories</h2>
        <style>
            #sidenav .accordion-body a {
                color : gray;
            }

            .accordion-body li {
                list-style : none;
            }
        </style>
        <ul class="nav nav-pills nav-stacked" id="sidenav">
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#QueueScripts">Queue Scripts</a>

                    <div id="QueueScripts"
                    <?php echo ($group == 'Queue Scripts')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVDiscovery.html">SVDiscovery</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVGenotyper.html">SVGenotyper</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVPreprocess.html">SVPreprocess</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVQScript.html">SVQScript</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotation">Variant Annotation</a>

                    <div id="VariantAnnotation"
                    <?php echo ($group == 'Variant Annotation')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_SVVariantAnnotatorWalker.html">SVVariantAnnotatorWalker</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotators">Variant Annotators</a>

                    <div id="VariantAnnotators"
                    <?php echo ($group == 'Variant Annotators')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_AlleleFrequencyAnnotator.html">AlleleFrequencyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ClusterSeparationAnnotator.html">ClusterSeparationAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_CopyNumberClassAnnotator.html">CopyNumberClassAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GCContentAnnotator.html">GCContentAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GeneOverlapAnnotator.html">GeneOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_IntensityRankSumAnnotator.html">IntensityRankSumAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_NonVariantAnnotator.html">NonVariantAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ReciprocalOverlapAnnotator.html">ReciprocalOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_RedundancyAnnotator.html">RedundancyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_TagVariantsAnnotator.html">TagVariantsAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_VariantsPerSampleAnnotator.html">VariantsPerSampleAnnotator</a>
                                </li>
                    </ul>
                    </div>
                </li>
        </ul>
    </aside>
    <?php getForumPosts( 'IntensityRankSumAnnotator' ) ?>

</section>

<div class="span8">

    <h1>IntensityRankSumAnnotator</h1>

    <p class="lead">Annotator for performing "in silico" validation of copy number variants using SNP array intensity data.</p>
        <h3>Category
            <small> Variant Annotators</small>
        </h3>
    <br/>

    <h2>Introduction</h2>
    

 <p>
 The IntensityRankSum annotator is invoked through the <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a> framework, which defines arguments common to all annotators.
 </p><p>
 The IntensityRankSum annotator uses SNP array intensity data to do a form of "in silico" validation of copy number variants.
 The general approach is to use a Wilcoxon rank sum test using the intensity data and to calculate a p-value as follows:
 For each array probe underlying the variant, each sample is assigned an integral rank for that probe.
 Then the set of ranks (across all probes) is combined and treated as a set of observations for the Wilcoxon rank sum test.
 If there is more than one probe, there will certainly be ties (i.e. some sample will be rank 1 with respect to each probe).
 Ties are broken randomly to assign the final ranking.  The random seed is based on the input data, so identical inputs will
 always produce identical results.  The Wilcoxon rank sum test is applied to the final combined ranking to test whether the
 event-carrying samples (see below) are shifted with respect to the non-event-carrying samples.  This is a one-tailed test
 of a negative shift (for deletions) or a positive shift (for duplications).
 </p><p>
 For CNVs, two tests are performed (and two p-values generated).  First, the appropriate reference ploidy is determined.
 The test is only performed on intervals of uniform gender-independent ploidy, or as special case on intervals of
 uniform gender-dependent ploidy if gender information is available and all of the input samples share the same gender.
 The two tests are a negative-shift test for samples with copy number less than the reference ploidy (typically two, i.e. diploid)
 compared to all samples whose copy number matches the reference ploidy and a positive-shift test for samples with copy
 number greater than the reference ploidy (typically two) compared to all samples whose copy number matches the reference ploidy.
 </p><p>
 This annotator can be run in one of several modes, depending on whether you have genotypes for each sample or simply a list
 of samples believed to carry the variant (either as homozygotes or heterozygotes).  The -irsUseGenotypes flag indicates which
 mode to use.
 </p><p>
 If you do not use genotypes (i.e. you are evaluating a set of discovered sites without genotypes), then the annotator will
 expect an INFO tag on each variant to indicate the set of samples that are thought to carry the variant (either as homozygotes or heterozygotes).
 The variants must be either deletions (SVTYPE=DEL) or duplications (SVTYPE=DUP).
 In addition, you must supply a list of samples that were evaluated during discovery (using -sample) so that the list of non-carrier samples can be determined for each variant.
 </p><p>
 If you use genotypes, then the annotator will process any kind of CNV (SVTYPE=DEL, SVTYPE=DUP or SVTYPE=CNV).
 Genotypes will be taken from the GT/GQ fields (if present) or from the CN/CNQ fields and the genotypes will be used to determine carrier and non-carrier samples.
 For CNVs, if GT/GQ fields are present then the alleles must be symbolic copy number alleles (e.g. &lt;CNx&gt;).
 </p>
 <h3>Input File Formats</h3>
 <p>
 The array intensity input file is a tab-delimited text file consisting of four fixed columns (ID, CHROM, START, END)
 and a variable number of additional columns, one per sample that has array data.
 The file consists of a header line specifying these four columns and then the name of each sample.
 After the header line, the file contains one line per array-probe location and must be sorted in reference-sequence order.
 Each array-probe location should have a single intensity value.  For SNP probes, this is usually the sum of the A and B normalized probe intensities.
 The START and END coordinates are generally the same, corresponding to the SNP position.
 </p>
 
 <h3>Output Formats</h3>
 <p>
 This annotator can produce the following outputs: Annotated output VCF, report file.
 </p><p>
 This annotator produces the following INFO field annotations for each VCF record (or the corresponding column in the report file):
 </p>
 <dl>
 <dt>IRS_NPROBES (NPROBES)</dt><dd>The number of array probes used for this variant.</dd>
 <dt>IRS_NSAMPLES (NSAMPLES)</dt><dd>The number of measured samples (either affected or unaffected).</dd>
 <dt>IRS_LOWERNSAMPLES (LOWERNSAMPLES)</dt><dd>The number of measured samples for the deletion p-value.</dd>
 <dt>IRS_LOWERPVALUE (LOWERPVALUE)</dt><dd>The p-value based on testing deleted samples (carrier samples with copy number less than two).</dd>
 <dt>IRS_UPPERNSAMPLES (UPPERNSAMPLES)</dt><dd>The number of measured samples for the duplication p-value.</dd>
 <dt>IRS_UPPERPVALUE (UPPERPVALUE)</dt><dd>The p-value based on testing duplicated samples (carrier samples with copy number greater than two).</dd>
 <dt>IRS_PVALUE (PVALUE)</dt><dd>The lesser of IRS_LOWERPVALUE and IRS_UPPERPVALUE.
   You should <i>not</i> use this minimum p-value for FDR estimation unless you compute and apply a suitable null expectation.
   Instead, you should stratify your variants based on type (deletion, duplication, mixed) and estimate FDR separately using IRS_LOWERPVALUE and IRS_UPPERPVALUE.</dd>
 </dl>
 
 <h3>Interpretation</h3>
 <p>
 The IntensityRankSum annotator provides basic statistics for each site, but the most robust way to interpret the results is to estimate the false discovery rate (FDR)
 of a large call set as a whole using the following procedure:  For each IRS p-value (LOWER or UPPER), calculate the ratio of the number of sites with a p-value >= 0.5
 to the number of sites with a valid p-value (i.e. sites with probes) and double this ratio to estimate the FDR of the call set as a whole.  This callset-wide estimation
 is more robust than trying to assess individual sites, as the p-values on individual sites are not always well-calibrated due to inflation in the tails of the p-value
 distribution.  As a result, using a p-value cutoff of 0.01 may yield a call set with a true FDR higher than 1%.
 </p>
 
 <h3>Examples</h3>
 <p>
 Example for assessing discovery, using INFO tags to determine carrier and non-carrier samples.
 </p><p>
 <pre>
 # Discovery example (using INFO tags)
 java -Xmx4g -cp SVToolkit.jar \
     org.broadinstitute.sv.main.SVAnnotator \
     -A IntensityRankSum \
     -R human_g1k_v37.fasta \
     -vcf input.vcf \
     -O output.vcf \
     -arrayIntensityFile ALL.genome.Omni25_probe_intensity_matrix.20110425.dat \
     -sample discovery_samples.list \
     -irsSampleTag IRSSAMPLES \
     -writeReport true \
     -reportFile irs_output.report.dat
 </pre>
 <p>
 Example for assessing genotyped variants, using the genotypes in the input VCF file to determine carrier and non-carrier samples.
 </p><p>
 <pre>
 # Genotyped example (using genotype fields)
 java -Xmx4g -cp SVToolkit.jar \
     org.broadinstitute.sv.main.SVAnnotator \
     -A IntensityRankSum \
     -R human_g1k_v37.fasta \
     -vcf input.vcf \
     -O output.vcf \
     -arrayIntensityFile ALL.genome.Omni25_probe_intensity_matrix.20110425.dat \
     -irsUseGenotypes true \
     -writeReport true \
     -reportFile irs_output.report.dat
 </pre>
 </p>

            <hr>
            <h3>IntensityRankSumAnnotator specific arguments</h3>
            <table class="table table-striped table-bordered table-condensed">
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Type</th>
                    <th>Default value</th>
                    <th>Summary</th>
                </tr>
                </thead>
                <tbody>
        <tr>
            <th colspan="4" id="row-divider">Required</th>
        </tr>
            <tr>
                <td><a href="#--arrayIntensityFile">--arrayIntensityFile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Path to file containing matrix of array intensity values</td>
            </tr>
        <tr>
            <th colspan="4" id="row-divider">Optional</th>
        </tr>
            <tr>
                <td><a href="#--genotypeQualityThreshold">--genotypeQualityThreshold</a></td>
                <td>Double</td>
                <td>NA</td>
                <td>When using genotypes, ignore genotypes below this genotype quality GQ value (default no threshold)</td>
            </tr>
            <tr>
                <td><a href="#--irsBoundaryPadding">--irsBoundaryPadding</a></td>
                <td>Integer</td>
                <td>0</td>
                <td>Probes within this distance of inner site boundaries are not used</td>
            </tr>
            <tr>
                <td><a href="#--irsPermute">--irsPermute</a></td>
                <td>String</td>
                <td>false</td>
                <td>Whether to permute samples to generate null distribution</td>
            </tr>
            <tr>
                <td><a href="#--irsSampleTag">--irsSampleTag</a></td>
                <td>String</td>
                <td>IRSSAMPLES</td>
                <td>VCF INFO tag containing list of carrier samples</td>
            </tr>
            <tr>
                <td><a href="#--irsUseGenotypes">--irsUseGenotypes</a></td>
                <td>String</td>
                <td>false</td>
                <td>Whether to use genotypes or INFO tags to select carrier samples</td>
            </tr>
            <tr>
                <td><a href="#--sample">--sample</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Sample(s) or .list file of sample names.  Used to determine the non-carrier samples when not using genotypes.</td>
            </tr>
                </tbody>
            </table>



                        <hr>
                            <h2>Argument details</h2>
    <h3><a name="--arrayIntensityFile">--arrayIntensityFile
         / -arrayIntensityFile
    </a>
        (
        required
        File
        
        )
    </h3>
    <p class="args">
        Path to file containing matrix of array intensity values. 
            </p>
    <h3><a name="--genotypeQualityThreshold">--genotypeQualityThreshold
         / -genotypeQualityThreshold
    </a>
        (
        
        Double
        
        )
    </h3>
    <p class="args">
        When using genotypes, ignore genotypes below this genotype quality GQ value (default no threshold). 
            </p>
    <h3><a name="--irsBoundaryPadding">--irsBoundaryPadding
         / -irsBoundaryPadding
    </a>
        (
        
        Integer
         with default value 0
        )
    </h3>
    <p class="args">
        Probes within this distance of inner site boundaries are not used. 
            </p>
    <h3><a name="--irsPermute">--irsPermute
         / -irsPermute
    </a>
        (
        
        String
         with default value false
        )
    </h3>
    <p class="args">
        Whether to permute samples to generate null distribution. 
            </p>
    <h3><a name="--irsSampleTag">--irsSampleTag
         / -irsSampleTag
    </a>
        (
        
        String
         with default value IRSSAMPLES
        )
    </h3>
    <p class="args">
        VCF INFO tag containing list of carrier samples. 
            </p>
    <h3><a name="--irsUseGenotypes">--irsUseGenotypes
         / -irsUseGenotypes
    </a>
        (
        
        String
         with default value false
        )
    </h3>
    <p class="args">
        Whether to use genotypes or INFO tags to select carrier samples. 
            </p>
    <h3><a name="--sample">--sample
         / -sample
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Sample(s) or .list file of sample names.  Used to determine the non-carrier samples when not using genotypes.. 
            </p>

        <hr>
        <p class="see-also">See also <a href="index.html">Documentation index</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip">GenomeSTRiP Site</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip/documentation">GenomeSTRiP Support</a></p>

        <p class="version">GenomeSTRiP version 1.04-r1441 built at 2014/07/14 14:26:06.</p>
        </div><!--/.span-->
        </div><!--/.row-->
        </body>
        <?php printFooter();
        ob_end_flush();
        ?>
