<!--
  ~ Copyright (c) 2013, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->

<!--
  ~ Copyright (c) 2012, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->



<?php $group = 'Variant Annotators'; ?>
        <?php ob_start(); ?>
        <!DOCTYPE html>
        <html lang='en'>
        <head>
            <meta charset='utf-8'>
            <title>GeneOverlapAnnotator documentation documentation</title>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <meta name='description' content=''>

            <link type="text/css" rel="stylesheet" href="bootstrap.min.css">
            <script src="jquery.min.js"></script>
            <script src="bootstrap.min.js"></script>
            <style>
                @media (min-width: 980px) {
                    body {
                        padding-top : 60px;
                    }
                }
            </style>

        </head>
        <body>
        <div class="navbar navbar-fixed-top">
            <div class="navbar-inner">
                <div class="container-fluid">
                    <a class="brand" href="index.html">GenomeSTRiP Documentation</a>
                </div>
            </div>
        </div>
        <div class="container-fluid">
            <?php
    // Output Buffering allows for the website to inject a different header
    ob_clean();
    include '../include/common.php';
    printHeader("GeneOverlapAnnotator documentation", "Guides");
    ?>

            <div class="row-fluid">

<section class="span4">
    <aside class="well">
        <h2>Categories</h2>
        <style>
            #sidenav .accordion-body a {
                color : gray;
            }

            .accordion-body li {
                list-style : none;
            }
        </style>
        <ul class="nav nav-pills nav-stacked" id="sidenav">
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#QueueScripts">Queue Scripts</a>

                    <div id="QueueScripts"
                    <?php echo ($group == 'Queue Scripts')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVDiscovery.html">SVDiscovery</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVGenotyper.html">SVGenotyper</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVPreprocess.html">SVPreprocess</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVQScript.html">SVQScript</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotation">Variant Annotation</a>

                    <div id="VariantAnnotation"
                    <?php echo ($group == 'Variant Annotation')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_SVVariantAnnotatorWalker.html">SVVariantAnnotatorWalker</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotators">Variant Annotators</a>

                    <div id="VariantAnnotators"
                    <?php echo ($group == 'Variant Annotators')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_AlleleFrequencyAnnotator.html">AlleleFrequencyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ClusterSeparationAnnotator.html">ClusterSeparationAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_CopyNumberClassAnnotator.html">CopyNumberClassAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GCContentAnnotator.html">GCContentAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GeneOverlapAnnotator.html">GeneOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_IntensityRankSumAnnotator.html">IntensityRankSumAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_NonVariantAnnotator.html">NonVariantAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ReciprocalOverlapAnnotator.html">ReciprocalOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_RedundancyAnnotator.html">RedundancyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_TagVariantsAnnotator.html">TagVariantsAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_VariantsPerSampleAnnotator.html">VariantsPerSampleAnnotator</a>
                                </li>
                    </ul>
                    </div>
                </li>
        </ul>
    </aside>
    <?php getForumPosts( 'GeneOverlapAnnotator' ) ?>

</section>

<div class="span8">

    <h1>GeneOverlapAnnotator</h1>

    <p class="lead">Annotator used to detect and report overlaps between structural variations and annotated genes.</p>
        <h3>Category
            <small> Variant Annotators</small>
        </h3>
    <br/>

    <h2>Introduction</h2>
    

 <p>
 The GeneOverlap annotator is invoked through the <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a> framework, which defines arguments common to all annotators.
 </p><p>
 The GeneOverlap annotator reports on the genes and transcripts that overlap a structural variation and the significance of the overlap (currently coding vs. exonic vs. intronic).
 </p>
 
 <h3>Input Formats</h3>
 <p>
 The -geneTrackFile must be specified in GTF format and certain restrictions on content and sort order must be followed.
 See the <a href="org_broadinstitute_sv_apps_SortGTFFile.html">SortGTFFile</a> utility program for details.
 </p><p>
 This annotator is new and so not all GTF files may work correctly.
 The GENCODE v17 GTF file works after running it through SortGTFFile.
 </p>
 <p></p>
 <p>
 The -geneDescriptionFile is a tab delimited file with two columns:  GENENAME and DESCRIPTION.
 The GENENAME column must match the gene names in the geneTrack GTF file.
 </p>
 
 <h3>Output Formats</h3>
 <p>
 This annotator can produce the following outputs: report file.
 </p><p>
 The report file contains one or more lines per input site.  If the input site overlaps one or more genes, one line is produced for each overlapping gene.
 </p><p>
 The report file contains the following columns:
 </p>
 <dl>
 <dt>ID</dt><dd>The variant ID from the input file.</dd>
 <dt>CHROM</dt><dd>The chromosome of the variant.</dd>
 <dt>START</dt><dd>The start coordinate of the variant on the reference.</dd>
 <dt>END</dt><dd>The end coordinate of the variant on the reference.</dd>
 <dt>NOVERLAP</dt><dd>The number of overlapping genes.</dd>
 <dt>GENENAME</dt><dd>The name of one overlapping gene.</dd>
 <dt>TRANSCRIPT</dt><dd>The transcript ID of the associated transcript or NA.</dd>
 <dt>GENEOVERLAP</dt><dd>The kind of overlap, currently GENE, CDS, EXON, INTRON, OTHER (or NA for no genic overlap).</dd>
 <dt>TXOVERLAP</dt><dd>The kind of transcript overlap, currently TX-ALL, TX-ALL-CDS, TX-EXON-FRAMESHIFT, TX-EXON-INFRAME, TX-TRUNCATED-3P, TX-TRUNCATED-5P, TX-CDS (or NA for no overlap).</dd>
 <dt>DESCRIPTION</dt><dd>A description of the gene, taken from the gene description file if supplied.</dd>
 </dl>
 
 <h3>Example</h3>
 <p>
 <pre>
 java -Xmx4g -cp SVToolkit.jar \
     org.broadinstitute.sv.main.SVAnnotator \
     -A GeneOverlap \
     -R human_g1k_v37.fasta \
     -vcf input.vcf \
     -geneTrackFile gencode.v17.annotation.sorted.gtf \
     -geneTypeFilter protein_coding \
     -geneDescriptionFile ucsc_gene_descriptions.dat \
     -writeReport true \
     -reportDirectory reportdir
 </pre>
 </p>

            <hr>
            <h3>GeneOverlapAnnotator specific arguments</h3>
            <table class="table table-striped table-bordered table-condensed">
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Type</th>
                    <th>Default value</th>
                    <th>Summary</th>
                </tr>
                </thead>
                <tbody>
        <tr>
            <th colspan="4" id="row-divider">Optional</th>
        </tr>
            <tr>
                <td><a href="#--geneDescriptionFile">--geneDescriptionFile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Tab-delimited two column file mapping gene names to text gene descriptions</td>
            </tr>
            <tr>
                <td><a href="#--geneTrackFile">--geneTrackFile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>File of gene models for gene overlap analysis (GTF format)</td>
            </tr>
            <tr>
                <td><a href="#--geneTypeFilter">--geneTypeFilter</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Set of gene types to consider when analyzing gene overlap</td>
            </tr>
            <tr>
                <td><a href="#--sequenceTranslationMapFile">--sequenceTranslationMapFile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Mapping file from sequence name aliases used in annotation files to reference sequence names (e.g. "chr1" to "1")</td>
            </tr>
                </tbody>
            </table>



                        <hr>
                            <h2>Argument details</h2>
    <h3><a name="--geneDescriptionFile">--geneDescriptionFile
         / -geneDescriptionFile
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Tab-delimited two column file mapping gene names to text gene descriptions. 
            </p>
    <h3><a name="--geneTrackFile">--geneTrackFile
         / -geneTrackFile
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        File of gene models for gene overlap analysis (GTF format). 
            </p>
    <h3><a name="--geneTypeFilter">--geneTypeFilter
         / -geneTypeFilter
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Set of gene types to consider when analyzing gene overlap. 
            </p>
    <h3><a name="--sequenceTranslationMapFile">--sequenceTranslationMapFile
         / -sequenceTranslationMapFile
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Mapping file from sequence name aliases used in annotation files to reference sequence names (e.g. "chr1" to "1"). 
            </p>

        <hr>
        <p class="see-also">See also <a href="index.html">Documentation index</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip">GenomeSTRiP Site</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip/documentation">GenomeSTRiP Support</a></p>

        <p class="version">GenomeSTRiP version 1.04-r1441 built at 2014/07/14 14:26:06.</p>
        </div><!--/.span-->
        </div><!--/.row-->
        </body>
        <?php printFooter();
        ob_end_flush();
        ?>
