<!--
  ~ Copyright (c) 2013, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->

<!--
  ~ Copyright (c) 2012, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->



<?php $group = 'Variant Annotators'; ?>
        <?php ob_start(); ?>
        <!DOCTYPE html>
        <html lang='en'>
        <head>
            <meta charset='utf-8'>
            <title>CopyNumberClassAnnotator documentation documentation</title>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <meta name='description' content=''>

            <link type="text/css" rel="stylesheet" href="bootstrap.min.css">
            <script src="jquery.min.js"></script>
            <script src="bootstrap.min.js"></script>
            <style>
                @media (min-width: 980px) {
                    body {
                        padding-top : 60px;
                    }
                }
            </style>

        </head>
        <body>
        <div class="navbar navbar-fixed-top">
            <div class="navbar-inner">
                <div class="container-fluid">
                    <a class="brand" href="index.html">GenomeSTRiP Documentation</a>
                </div>
            </div>
        </div>
        <div class="container-fluid">
            <?php
    // Output Buffering allows for the website to inject a different header
    ob_clean();
    include '../include/common.php';
    printHeader("CopyNumberClassAnnotator documentation", "Guides");
    ?>

            <div class="row-fluid">

<section class="span4">
    <aside class="well">
        <h2>Categories</h2>
        <style>
            #sidenav .accordion-body a {
                color : gray;
            }

            .accordion-body li {
                list-style : none;
            }
        </style>
        <ul class="nav nav-pills nav-stacked" id="sidenav">
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#QueueScripts">Queue Scripts</a>

                    <div id="QueueScripts"
                    <?php echo ($group == 'Queue Scripts')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVDiscovery.html">SVDiscovery</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVGenotyper.html">SVGenotyper</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVPreprocess.html">SVPreprocess</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVQScript.html">SVQScript</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotation">Variant Annotation</a>

                    <div id="VariantAnnotation"
                    <?php echo ($group == 'Variant Annotation')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_SVVariantAnnotatorWalker.html">SVVariantAnnotatorWalker</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotators">Variant Annotators</a>

                    <div id="VariantAnnotators"
                    <?php echo ($group == 'Variant Annotators')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_AlleleFrequencyAnnotator.html">AlleleFrequencyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ClusterSeparationAnnotator.html">ClusterSeparationAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_CopyNumberClassAnnotator.html">CopyNumberClassAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GCContentAnnotator.html">GCContentAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GeneOverlapAnnotator.html">GeneOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_IntensityRankSumAnnotator.html">IntensityRankSumAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_NonVariantAnnotator.html">NonVariantAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ReciprocalOverlapAnnotator.html">ReciprocalOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_RedundancyAnnotator.html">RedundancyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_TagVariantsAnnotator.html">TagVariantsAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_VariantsPerSampleAnnotator.html">VariantsPerSampleAnnotator</a>
                                </li>
                    </ul>
                    </div>
                </li>
        </ul>
    </aside>
    <?php getForumPosts( 'CopyNumberClassAnnotator' ) ?>

</section>

<div class="span8">

    <h1>CopyNumberClassAnnotator</h1>

    <p class="lead">Annotator that report statistics about the distribution of copy number states at each variant.</p>
        <h3>Category
            <small> Variant Annotators</small>
        </h3>
    <br/>

    <h2>Introduction</h2>
    

 <p>
 The CopyNumberClass annotator is invoked through the <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a> framework, which defines arguments common to all annotators.
 </p><p>
 The CopyNumberClass annotator reports several statistics about the distribution of copy number states across the genotyped samples.
 The -classifyMinimumObservations argument controls how many times a copy number state must be (confidently) seen to count that copy number state as being "observed" in the population.
 The default is 1, meaning any confident genotype call will cause the copy number state to be observed. All of the outputs (except call rate) only consider confident genotypes (based
 on the filters and quality values) in observed copy number states. Unobserved copy number states are treated like no-calls.
 </p><p>
 This annotator is gender and ploidy aware. The copy number distribution includes all samples (independent of gender), but the calculation of the minimum number of requried alleles
 is gender-dependent, as is the determination of whether a sample must carry a non-reference allele (NNONREF) or how many samples differ from the mode copy number state (NVARIANT).
 NVARIANT is a measure of how variable a site is, independent of the copy number of the reference allele. NVARIANT is the number of samples that are different from the most-common
 (mode) copy number state. For sex chromosomes, this is calcualted independently for males and females and then summed.
 </p>

 <h3>Output Formats</h3>
 <p>
 This annotator can produce the following outputs: report file.
 </p><p>
 The report file contains one line per variant, with the following columns:
 </p>
 <dl>
 <dt>ID</dt><dd>The variant identifier.</dd>
 <dt>CALLRATE</dt><dd>The genotyping call rate for this site.</dd>
 <dt>CNMIN</dt><dd>The minimum observed copy number class.</dd>
 <dt>CNMAX</dt><dd>The maximum observed copy number class.</dd>
 <dt>CNALLELES</dt><dd>The minimum number of haploid copy number alleles that would be required to explain the observed copy number states. This calculation is gender-aware.</dd>
 <dt>NNONREF</dt><dd>The number of samples that are called different than uniformly reference. This calculation is gender-aware.</dd>
 <dt>NVARIANT</dt><dd>The number of samples that are called different than the mode copy number. For sex chromosomes, this is calculated independently for males and females and summed.</dd>
 <dt>CNDIST</dt><dd>A comma-separated list of integers given the observed number of samples in each copy number state, starting from zero. Males and females are combined.</dd>
 </dl>
 
 <h3>Example</h3>
 <p>
 <pre>
 java -Xmx4g -cp SVToolkit.jar \
     org.broadinstitute.sv.main.SVAnnotator \
     -A CopyNumberClass \
     -R human_g1k_v37.fasta \
     -vcf input.vcf \
     -writeReport true \
     -reportDirectory reportdir
 </pre>
 </p>

            <hr>
            <h3>CopyNumberClassAnnotator specific arguments</h3>
            <table class="table table-striped table-bordered table-condensed">
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Type</th>
                    <th>Default value</th>
                    <th>Summary</th>
                </tr>
                </thead>
                <tbody>
        <tr>
            <th colspan="4" id="row-divider">Optional</th>
        </tr>
            <tr>
                <td><a href="#--classifyMinimumObservations">--classifyMinimumObservations</a></td>
                <td>Integer</td>
                <td>NA</td>
                <td>Minimum number of confident observations required when classifying copy number states (default 1)</td>
            </tr>
            <tr>
                <td><a href="#--filterGenotypes">--filterGenotypes</a></td>
                <td>String</td>
                <td>NA</td>
                <td>True to ignore genotypes that have been filtered (default true)</td>
            </tr>
            <tr>
                <td><a href="#--genotypeQualityThreshold">--genotypeQualityThreshold</a></td>
                <td>Double</td>
                <td>NA</td>
                <td>Ignore genotypes below this genotype quality GQ/CNQ value (default no threshold)</td>
            </tr>
            <tr>
                <td><a href="#--sample">--sample</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Sample(s) or .list file of sample names.  If specified, only the listed samples will be used to evaluate the variants.</td>
            </tr>
                </tbody>
            </table>



                        <hr>
                            <h2>Argument details</h2>
    <h3><a name="--classifyMinimumObservations">--classifyMinimumObservations
         / -classifyMinimumObservations
    </a>
        (
        
        Integer
        
        )
    </h3>
    <p class="args">
        Minimum number of confident observations required when classifying copy number states (default 1). 
            </p>
    <h3><a name="--filterGenotypes">--filterGenotypes
         / -filterGenotypes
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        True to ignore genotypes that have been filtered (default true). 
            </p>
    <h3><a name="--genotypeQualityThreshold">--genotypeQualityThreshold
         / -genotypeQualityThreshold
    </a>
        (
        
        Double
        
        )
    </h3>
    <p class="args">
        Ignore genotypes below this genotype quality GQ/CNQ value (default no threshold). 
            </p>
    <h3><a name="--sample">--sample
         / -sample
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Sample(s) or .list file of sample names.  If specified, only the listed samples will be used to evaluate the variants.. 
            </p>

        <hr>
        <p class="see-also">See also <a href="index.html">Documentation index</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip">GenomeSTRiP Site</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip/documentation">GenomeSTRiP Support</a></p>

        <p class="version">GenomeSTRiP version 1.04-r1441 built at 2014/07/14 14:26:06.</p>
        </div><!--/.span-->
        </div><!--/.row-->
        </body>
        <?php printFooter();
        ob_end_flush();
        ?>
