<!--
  ~ Copyright (c) 2013, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->

<!--
  ~ Copyright (c) 2012, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->



<?php $group = 'Variant Annotators'; ?>
        <?php ob_start(); ?>
        <!DOCTYPE html>
        <html lang='en'>
        <head>
            <meta charset='utf-8'>
            <title>ReciprocalOverlapAnnotator documentation documentation</title>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <meta name='description' content=''>

            <link type="text/css" rel="stylesheet" href="bootstrap.min.css">
            <script src="jquery.min.js"></script>
            <script src="bootstrap.min.js"></script>
            <style>
                @media (min-width: 980px) {
                    body {
                        padding-top : 60px;
                    }
                }
            </style>

        </head>
        <body>
        <div class="navbar navbar-fixed-top">
            <div class="navbar-inner">
                <div class="container-fluid">
                    <a class="brand" href="index.html">GenomeSTRiP Documentation</a>
                </div>
            </div>
        </div>
        <div class="container-fluid">
            <?php
    // Output Buffering allows for the website to inject a different header
    ob_clean();
    include '../include/common.php';
    printHeader("ReciprocalOverlapAnnotator documentation", "Guides");
    ?>

            <div class="row-fluid">

<section class="span4">
    <aside class="well">
        <h2>Categories</h2>
        <style>
            #sidenav .accordion-body a {
                color : gray;
            }

            .accordion-body li {
                list-style : none;
            }
        </style>
        <ul class="nav nav-pills nav-stacked" id="sidenav">
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#QueueScripts">Queue Scripts</a>

                    <div id="QueueScripts"
                    <?php echo ($group == 'Queue Scripts')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVDiscovery.html">SVDiscovery</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVGenotyper.html">SVGenotyper</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVPreprocess.html">SVPreprocess</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVQScript.html">SVQScript</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotation">Variant Annotation</a>

                    <div id="VariantAnnotation"
                    <?php echo ($group == 'Variant Annotation')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_SVVariantAnnotatorWalker.html">SVVariantAnnotatorWalker</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotators">Variant Annotators</a>

                    <div id="VariantAnnotators"
                    <?php echo ($group == 'Variant Annotators')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_AlleleFrequencyAnnotator.html">AlleleFrequencyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ClusterSeparationAnnotator.html">ClusterSeparationAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_CopyNumberClassAnnotator.html">CopyNumberClassAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GCContentAnnotator.html">GCContentAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GeneOverlapAnnotator.html">GeneOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_IntensityRankSumAnnotator.html">IntensityRankSumAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_NonVariantAnnotator.html">NonVariantAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ReciprocalOverlapAnnotator.html">ReciprocalOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_RedundancyAnnotator.html">RedundancyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_TagVariantsAnnotator.html">TagVariantsAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_VariantsPerSampleAnnotator.html">VariantsPerSampleAnnotator</a>
                                </li>
                    </ul>
                    </div>
                </li>
        </ul>
    </aside>
    <?php getForumPosts( 'ReciprocalOverlapAnnotator' ) ?>

</section>

<div class="span8">

    <h1>ReciprocalOverlapAnnotator</h1>

    <p class="lead">Annotator for computing reciprocal overlap between sets of structural variants.</p>
        <h3>Category
            <small> Variant Annotators</small>
        </h3>
    <br/>

    <h2>Introduction</h2>
    

 <p>
 The ReciprocalOverlap annotator is invoked through the <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a> framework, which defines arguments common to all annotators.
 </p><p>
 The ReciprocalOverlap annotator computes the overlap between sets of structural variants, represented as intervals.
 Compared to other tools, some useful features of the reciprocal overlap annotator are that it processes VCF files, it understands VCF confidence intervals and it supports multiple
 modes for selecting the "best" overlapping intervals.
 </p><p>
 This annotator compares each variant in the input VCF to all variants in a second (comparison) VCF.  The comparison VCF can be the same file as the input VCF to compute
 the self-overlap of a set of variants.
 </p><p>
 The intervals used for comparison are based on the -variantIntervalMode argument and are different if confidence intervals are specified in the VCF files.
 An interval mode of MINIMUM uses the smallest (innermost) reference interval for each variant, MAXIMUM uses the largest (outermost) reference interval and NOMINAL (the default)
 uses the exact POS/END values as specified in the VCF.
 </p><p>
 For each input variant, a "best hit" comparison variant is chosen among all of the overlapping variants.  The best hit is selected either as the variant with the greatest
 overlap fraction (the default) or the greatest overlap length in bases, determined by the -reciprocalOverlapRankBy argument.
 </p><p>
 By default, variants with the same ID are not compared (to allow easy computation of the self-overlap of a call set).  This can be disabled with -reciprocalOverlapCompareIds false,
 which will force variants with the same ID to be compared.
 </p>
 
 <h3>Output Formats</h3>
 <p>
 This annotator can produce the following outputs: annotated VCF, report file.
 </p><p>
 The following VCF annotations are produced by this annotator:
 </p>
 <dl>
 <dt>RO_NHITS</dt><dd>The number of comparison variants that overlap this input variant.</dd>
 <dt>RO_BESTHIT</dt><dd>The identifier of the comparison variant with the greatest overlap to this variant.</dd>
 <dt>RO_LENGTH</dt><dd>The length of this variant as used in the overlap calculation.</dd>
 <dt>RO_BHLENGTH</dt><dd>The length of the best hit variant as used in the overlap calculation.</dd>
 <dt>RO_OVERLAP</dt><dd>The overlap fraction between this variant and the best hit.</dd>
 <dt>RO_OVERLAPLENTH</dt><dd>The length of the overlap between this variant and the best hit.</dd>
 </dl>
 </p><p>
 The report file contains one line for every input variant. The following columns are produced in the report file:
 </p>
 <dl>
 <dt>VARIANT</dt><dd>The ID of the input variant.</dd>
 <dt>CHROM</dt><dd>The chromosome of the input variant.</dd>
 <dt>START</dt><dd>The start coordinate of the input variant on the reference.</dd>
 <dt>END</dt><dd>The end coordinate of the input variant on the reference.</dd>
 <dt>NOVERLAP</dt><dd>The number of overlapping comparison variants.</dd>
 <dt>BESTHIT</dt><dd>The identifier of the comparison variant with the greatest overlap to this variant.</dd>
 <dt>BHSTART</dt><dd>The start coordinate of the best hit variant on the reference.</dd>
 <dt>BHEND</dt><dd>The end coordinate of the best hit variant on the reference.</dd>
 <dt>LENGTH</dt><dd>The length of the input variant as used in the overlap calculation.</dd>
 <dt>BHLENGTH</dt><dd>The length of the best hit variant as used in the overlap calculation.</dd>
 <dt>UNIONLENGTH</dt><dd>The length of the union of the input and best hit variant.</dd>
 <dt>OVERLAPLENTH</dt><dd>The length of the overlap (intersection) between this variant and the best hit variant.</dd>
 <dt>OVERLAP</dt><dd>The overlap fraction between this variant and the best hit.</dd>
 </dl>
 
 <h3>Example</h3>
 <p>
 <pre>
 java -Xmx4g -cp SVToolkit.jar \
     org.broadinstitute.sv.main.SVAnnotator \
     -A ReciprocalOverlap \
     -R human_g1k_v37.fasta \
     -vcf input.vcf \
     -comparisonFile comparison.vcf \
     -O output.vcf \
     -writeReport true \
     -reportDirectory reportdir
 </pre>
 </p>

            <hr>
            <h3>ReciprocalOverlapAnnotator specific arguments</h3>
            <table class="table table-striped table-bordered table-condensed">
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Type</th>
                    <th>Default value</th>
                    <th>Summary</th>
                </tr>
                </thead>
                <tbody>
        <tr>
            <th colspan="4" id="row-divider">Required</th>
        </tr>
            <tr>
                <td><a href="#--comparisonFile">--comparisonFile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>VCF file of variants for comparison</td>
            </tr>
        <tr>
            <th colspan="4" id="row-divider">Optional</th>
        </tr>
            <tr>
                <td><a href="#--filterVariants">--filterVariants</a></td>
                <td>String</td>
                <td>NA</td>
                <td>True to ignore variants that have been filtered (default true).</td>
            </tr>
            <tr>
                <td><a href="#--reciprocalOverlapCompareIds">--reciprocalOverlapCompareIds</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Use variant IDs to avoid comparing a variant with itself to allow self-overlap comparisons (default true)</td>
            </tr>
            <tr>
                <td><a href="#--reciprocalOverlapRankBy">--reciprocalOverlapRankBy</a></td>
                <td>String</td>
                <td>NA</td>
                <td>How to select between multiple overlapping variants, possible values: LENGTH, FRACTION (default)</td>
            </tr>
            <tr>
                <td><a href="#--variantIntervalMode">--variantIntervalMode</a></td>
                <td>String</td>
                <td>0</td>
                <td>How to measure variant intervals, possible values: NOMINAL (default), MINIMUM, MAXIMUM</td>
            </tr>
                </tbody>
            </table>



                        <hr>
                            <h2>Argument details</h2>
    <h3><a name="--comparisonFile">--comparisonFile
         / -comparisonFile
    </a>
        (
        required
        File
        
        )
    </h3>
    <p class="args">
        VCF file of variants for comparison. 
            </p>
    <h3><a name="--filterVariants">--filterVariants
         / -filterVariants
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        True to ignore variants that have been filtered (default true).. 
            </p>
    <h3><a name="--reciprocalOverlapCompareIds">--reciprocalOverlapCompareIds
         / -reciprocalOverlapCompareIds
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Use variant IDs to avoid comparing a variant with itself to allow self-overlap comparisons (default true). 
            </p>
    <h3><a name="--reciprocalOverlapRankBy">--reciprocalOverlapRankBy
         / -reciprocalOverlapRankBy
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        How to select between multiple overlapping variants, possible values: LENGTH, FRACTION (default). 
            </p>
    <h3><a name="--variantIntervalMode">--variantIntervalMode
         / -variantIntervalMode
    </a>
        (
        
        String
         with default value 0
        )
    </h3>
    <p class="args">
        How to measure variant intervals, possible values: NOMINAL (default), MINIMUM, MAXIMUM. 
            </p>

        <hr>
        <p class="see-also">See also <a href="index.html">Documentation index</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip">GenomeSTRiP Site</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip/documentation">GenomeSTRiP Support</a></p>

        <p class="version">GenomeSTRiP version 1.04-r1441 built at 2014/07/14 14:26:06.</p>
        </div><!--/.span-->
        </div><!--/.row-->
        </body>
        <?php printFooter();
        ob_end_flush();
        ?>
