<!--
  ~ Copyright (c) 2013, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->

<!--
  ~ Copyright (c) 2012, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->



<?php $group = 'Variant Annotation'; ?>
        <?php ob_start(); ?>
        <!DOCTYPE html>
        <html lang='en'>
        <head>
            <meta charset='utf-8'>
            <title>SVAnnotator documentation documentation</title>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <meta name='description' content=''>

            <link type="text/css" rel="stylesheet" href="bootstrap.min.css">
            <script src="jquery.min.js"></script>
            <script src="bootstrap.min.js"></script>
            <style>
                @media (min-width: 980px) {
                    body {
                        padding-top : 60px;
                    }
                }
            </style>

        </head>
        <body>
        <div class="navbar navbar-fixed-top">
            <div class="navbar-inner">
                <div class="container-fluid">
                    <a class="brand" href="index.html">GenomeSTRiP Documentation</a>
                </div>
            </div>
        </div>
        <div class="container-fluid">
            <?php
    // Output Buffering allows for the website to inject a different header
    ob_clean();
    include '../include/common.php';
    printHeader("SVAnnotator documentation", "Guides");
    ?>

            <div class="row-fluid">

<section class="span4">
    <aside class="well">
        <h2>Categories</h2>
        <style>
            #sidenav .accordion-body a {
                color : gray;
            }

            .accordion-body li {
                list-style : none;
            }
        </style>
        <ul class="nav nav-pills nav-stacked" id="sidenav">
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#QueueScripts">Queue Scripts</a>

                    <div id="QueueScripts"
                    <?php echo ($group == 'Queue Scripts')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVDiscovery.html">SVDiscovery</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVGenotyper.html">SVGenotyper</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVPreprocess.html">SVPreprocess</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVQScript.html">SVQScript</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotation">Variant Annotation</a>

                    <div id="VariantAnnotation"
                    <?php echo ($group == 'Variant Annotation')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_SVVariantAnnotatorWalker.html">SVVariantAnnotatorWalker</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotators">Variant Annotators</a>

                    <div id="VariantAnnotators"
                    <?php echo ($group == 'Variant Annotators')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_AlleleFrequencyAnnotator.html">AlleleFrequencyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ClusterSeparationAnnotator.html">ClusterSeparationAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_CopyNumberClassAnnotator.html">CopyNumberClassAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GCContentAnnotator.html">GCContentAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GeneOverlapAnnotator.html">GeneOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_IntensityRankSumAnnotator.html">IntensityRankSumAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_NonVariantAnnotator.html">NonVariantAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ReciprocalOverlapAnnotator.html">ReciprocalOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_RedundancyAnnotator.html">RedundancyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_TagVariantsAnnotator.html">TagVariantsAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_VariantsPerSampleAnnotator.html">VariantsPerSampleAnnotator</a>
                                </li>
                    </ul>
                    </div>
                </li>
        </ul>
    </aside>
    <?php getForumPosts( 'SVAnnotator' ) ?>

</section>

<div class="span8">

    <h1>SVAnnotator</h1>

    <p class="lead">Annotation engine and framework for SV annotation.</p>
        <h3>Category
            <small> Variant Annotation</small>
        </h3>
    <br/>

    <h2>Introduction</h2>
    

 <p>
 SVAnnotator is a general framework for generating annotations of different kinds on VCF files containing structural variant records.
 </p><p>
 SVAnnotator is conceptually similar to the GATK VariantAnnotator walker, but works somewhat differently and is tailored for use with structural variations.
 Each annotator is like a plug-in and you can run one or more annotators over the same file in one invocation.
 </p><p>
 Each annotator can either add annotations (typically INFO fields) to the input VCF file, or it can generate a textual report (e.g. tab delimited)
 for additional processing, or it can generate one or more summary reports (also text files) or all of the above.  Different annotators may or may
 not support all modes of operation. Consult the documentation for each individual annotator to see what output formats it supports.
 </p>
 <h3>Common Framework Arguments</h3>
 <p>
 The SVAnnotator framework supports a number of generic command line arguments that are documented below.
 Each individual annotator may support additional arguments.
 </p><p>
 Where possible, annotators will try to use the same interpretation to the same arguments.
 For example, the -sample argument is supported by a number of annotators to specify a (restricted) set of samples to use in the analysis.
 Following standard SVToolkit conventions, this -sample argument can be present multiple times and can list either individual samples or a .list file
 (a text file with extension .list) that contains a list of sample identifiers, or a combination of both.
 </p><p>
 The only required arguments are a reference sequence (-R) and an input VCF file (-vcf).  The set of annotators to run is specified with -A.
 Individual annotators may require additional specific arguments.
 </p>
 
 <h3>Examples</h3>
 <p>
 Run several annotators, generating an annotated VCF file and also report and summary files (for annotators that support them).
 </p><p>
 <pre>
 java -Xmx4g -cp SVToolkit.jar:GenomeAnalysisTK.jar \
     org.broadinstitute.sv.main.SVAnnotator \
     -A GCContent \
     -A VarantsPerSample \
     -A NonVariant \
     -A Redundancy \
     -R human_g1k_v37.fasta \
     -vcf input.vcf \
     -O output.vcf \
     -comparisonFile input.vcf \  # required for Redundancy annotator
     -writeReport true \
     -writeSummary true \
     -reportDirectory outputDir
 </pre>

            <hr>
            <h3>SVAnnotator specific arguments</h3>
            <table class="table table-striped table-bordered table-condensed">
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Type</th>
                    <th>Default value</th>
                    <th>Summary</th>
                </tr>
                </thead>
                <tbody>
        <tr>
            <th colspan="4" id="row-divider">Required</th>
        </tr>
            <tr>
                <td><a href="#--reference_sequence">--reference_sequence</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Reference sequence (indexed fasta file)</td>
            </tr>
        <tr>
            <th colspan="4" id="row-divider">Optional</th>
        </tr>
            <tr>
                <td><a href="#--annotation">--annotation</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>One or more specific annotations to apply to variant calls</td>
            </tr>
            <tr>
                <td><a href="#--outputFile">--outputFile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>VCF file to which annotated variants should be written</td>
            </tr>
            <tr>
                <td><a href="#--parameter">--parameter</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Override individual configuration parameters (for advanced use only)</td>
            </tr>
            <tr>
                <td><a href="#--reportDirectory">--reportDirectory</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Directory for reports (default is current directory)</td>
            </tr>
            <tr>
                <td><a href="#--reportFile">--reportFile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>File name for report (default is based an annotator name)</td>
            </tr>
            <tr>
                <td><a href="#--reportFileMap">--reportFileMap</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Map of file name for reports, overrides reportFile argument when multiple annotators are run</td>
            </tr>
            <tr>
                <td><a href="#--summaryFile">--summaryFile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>File name for summary (default is based an annotator name)</td>
            </tr>
            <tr>
                <td><a href="#--summaryFileMap">--summaryFileMap</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Map of file name for summaries, overrides summaryFile argument when multiple annotators are run</td>
            </tr>
            <tr>
                <td><a href="#--tempDir">--tempDir</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Directory for temporary files</td>
            </tr>
            <tr>
                <td><a href="#--useGATKTraversal">--useGATKTraversal</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Use normal GATK walker traversal (default false)</td>
            </tr>
            <tr>
                <td><a href="#--vcfFile">--vcfFile</a></td>
                <td>File</td>
                <td>NA</td>
                <td>Input variant file (VCF format)</td>
            </tr>
            <tr>
                <td><a href="#--writeReport">--writeReport</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Whether to write a report (default false)</td>
            </tr>
            <tr>
                <td><a href="#--writeSummary">--writeSummary</a></td>
                <td>String</td>
                <td>NA</td>
                <td>Whether to write a summary (default false)</td>
            </tr>
                </tbody>
            </table>



                        <hr>
                            <h2>Argument details</h2>
    <h3><a name="--annotation">--annotation
         / -A
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        One or more specific annotations to apply to variant calls. 
            </p>
    <h3><a name="--outputFile">--outputFile
         / -O
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        VCF file to which annotated variants should be written. 
            </p>
    <h3><a name="--parameter">--parameter
         / -P
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Override individual configuration parameters (for advanced use only). 
            </p>
    <h3><a name="--reference_sequence">--reference_sequence
         / -R
    </a>
        (
        required
        File
        
        )
    </h3>
    <p class="args">
        Reference sequence (indexed fasta file). 
            </p>
    <h3><a name="--reportDirectory">--reportDirectory
         / -reportDirectory
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Directory for reports (default is current directory). 
            </p>
    <h3><a name="--reportFile">--reportFile
         / -reportFile
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        File name for report (default is based an annotator name). 
            </p>
    <h3><a name="--reportFileMap">--reportFileMap
         / -reportFileMap
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Map of file name for reports, overrides reportFile argument when multiple annotators are run. 
            </p>
    <h3><a name="--summaryFile">--summaryFile
         / -summaryFile
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        File name for summary (default is based an annotator name). 
            </p>
    <h3><a name="--summaryFileMap">--summaryFileMap
         / -summaryFileMap
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Map of file name for summaries, overrides summaryFile argument when multiple annotators are run. 
            </p>
    <h3><a name="--tempDir">--tempDir
         / -tempDir
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Directory for temporary files. 
            </p>
    <h3><a name="--useGATKTraversal">--useGATKTraversal
         / -useGATKTraversal
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Use normal GATK walker traversal (default false). 
            </p>
    <h3><a name="--vcfFile">--vcfFile
         / -vcf
    </a>
        (
        
        File
        
        )
    </h3>
    <p class="args">
        Input variant file (VCF format). 
            </p>
    <h3><a name="--writeReport">--writeReport
         / -writeReport
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Whether to write a report (default false). 
            </p>
    <h3><a name="--writeSummary">--writeSummary
         / -writeSummary
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        Whether to write a summary (default false). 
            </p>

        <hr>
        <p class="see-also">See also <a href="index.html">Documentation index</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip">GenomeSTRiP Site</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip/documentation">GenomeSTRiP Support</a></p>

        <p class="version">GenomeSTRiP version 1.04-r1441 built at 2014/07/14 14:26:06.</p>
        </div><!--/.span-->
        </div><!--/.row-->
        </body>
        <?php printFooter();
        ob_end_flush();
        ?>
