<!--
  ~ Copyright (c) 2013, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->

<!--
  ~ Copyright (c) 2012, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->



<?php $group = 'Variant Annotators'; ?>
        <?php ob_start(); ?>
        <!DOCTYPE html>
        <html lang='en'>
        <head>
            <meta charset='utf-8'>
            <title>TagVariantsAnnotator documentation documentation</title>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <meta name='description' content=''>

            <link type="text/css" rel="stylesheet" href="bootstrap.min.css">
            <script src="jquery.min.js"></script>
            <script src="bootstrap.min.js"></script>
            <style>
                @media (min-width: 980px) {
                    body {
                        padding-top : 60px;
                    }
                }
            </style>

        </head>
        <body>
        <div class="navbar navbar-fixed-top">
            <div class="navbar-inner">
                <div class="container-fluid">
                    <a class="brand" href="index.html">GenomeSTRiP Documentation</a>
                </div>
            </div>
        </div>
        <div class="container-fluid">
            <?php
    // Output Buffering allows for the website to inject a different header
    ob_clean();
    include '../include/common.php';
    printHeader("TagVariantsAnnotator documentation", "Guides");
    ?>

            <div class="row-fluid">

<section class="span4">
    <aside class="well">
        <h2>Categories</h2>
        <style>
            #sidenav .accordion-body a {
                color : gray;
            }

            .accordion-body li {
                list-style : none;
            }
        </style>
        <ul class="nav nav-pills nav-stacked" id="sidenav">
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#QueueScripts">Queue Scripts</a>

                    <div id="QueueScripts"
                    <?php echo ($group == 'Queue Scripts')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVDiscovery.html">SVDiscovery</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVGenotyper.html">SVGenotyper</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVPreprocess.html">SVPreprocess</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVQScript.html">SVQScript</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotation">Variant Annotation</a>

                    <div id="VariantAnnotation"
                    <?php echo ($group == 'Variant Annotation')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_SVVariantAnnotatorWalker.html">SVVariantAnnotatorWalker</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotators">Variant Annotators</a>

                    <div id="VariantAnnotators"
                    <?php echo ($group == 'Variant Annotators')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_AlleleFrequencyAnnotator.html">AlleleFrequencyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ClusterSeparationAnnotator.html">ClusterSeparationAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_CopyNumberClassAnnotator.html">CopyNumberClassAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GCContentAnnotator.html">GCContentAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GeneOverlapAnnotator.html">GeneOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_IntensityRankSumAnnotator.html">IntensityRankSumAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_NonVariantAnnotator.html">NonVariantAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ReciprocalOverlapAnnotator.html">ReciprocalOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_RedundancyAnnotator.html">RedundancyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_TagVariantsAnnotator.html">TagVariantsAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_VariantsPerSampleAnnotator.html">VariantsPerSampleAnnotator</a>
                                </li>
                    </ul>
                    </div>
                </li>
        </ul>
    </aside>
    <?php getForumPosts( 'TagVariantsAnnotator' ) ?>

</section>

<div class="span8">

    <h1>TagVariantsAnnotator</h1>

    <p class="lead">Annotator that evaluates tagging SNPs (or other markers) based on pairwise r<sup>2</sup>.</p>
        <h3>Category
            <small> Variant Annotators</small>
        </h3>
    <br/>

    <h2>Introduction</h2>
    

 <p>
 The TagVariants annotator is invoked through the <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a> framework, which defines arguments common to all annotators.
 </p>
 </p><p>
 The TagVariants annotator compares the input variant file to a comparison file of variants (e.g. VCF) specified by -tagFile, which can be the same file or a different file.
 For each variant in the input variant file, a pairwise r2 value is computed to all variants in the tag variant file within a certain genomic distance of each
 input variant (by default 1 megabase).
 </p><p>
 This annotator can process either bi-allelic variants (with GT/GQ fields) or copy-number variants (with CN/CNQ fields).
 In the case of copy number variants, r2 is computed with respect to diploid copy number.
 </p><p>
 Multiple populations (potentially overlapping) may be specified and then the r2 calculations are performed separately in each population.
 If no populations are defined, the samples in the VCF are treated as a single population.
 The r2 values are only calculated for samples that are present in both the input VCF file and the tag VCF file.
 </p><p>
 This annotator produces both a summary report and a detailed report.
 The detailed report contains r2 values for all pairs of vairants above a user-specified r2 threshold (see -tagR2Threshold).
 The summary report contains the best r2 value for each input marker in each population.
 </p><p>
 The detailed report file contains a line for each pair of variants and each population where the r2 value is above the specified threshold.
 If this threshold is low and the tag window size is large, the report file may contain many lines for each variant.
 </p>
 
 <h3>Input Formats</h3>
 <p>
 Population map files are tab delimited files with two columns.  The first column specifies the sample identifier and the second column
 specifies a population identifier.  A header line is optional, but if present the column names should be SAMPLE and POPULATION.
 </p><p>
 Multiple population map files may be provided and these may assign multiple populations to the same sample.  This allows multiple levels
 of population structure or overlapping populations to be described.
 </p>
 
 <h3>Output Formats</h3>
 <p>
 This annotator can produce the following outputs: summary file, report file.
 </p><p>
 The summary file contains one line for each input variant.  The populations are reported in separate columns.
 </p><p>
 The columns in the summary file are:
 </p>
 <dl>
 <dt>VARIANT</dt><dd>The ID of the variant.</dd>
 <dt>CHR</dt><dd>The chromosome of the variant.</dd>
 <dt>START</dt><dd>The start position of the variant.</dd>
 <dt>END</dt><dd>The end position of the variant.</dd>
 <dt>MAF</dt><dd>The minor allele frequency of the variant (only present when there is a single population).</dd>
 <dt>R2</dt><dd>The maximum r2 value found to the best tagging marker.
 This column is displayed as "R2" when there is only one population.
 If there are multiple populations being tested, then there will be one column for each population and the column header specifies the population identifier.
 </dd>
 </dl>
 <p>
 </p>
 The columns in the detailed report file are:
 </p>
 <dl>
 <dt>VARIANT</dt><dd>The ID of the input variant.</dd>
 <dt>CHR</dt><dd>The chromosome of the input variant.</dd>
 <dt>START</dt><dd>The start position of the input variant.</dd>
 <dt>END</dt><dd>The end position of the input variant.</dd>
 <dt>MAF</dt><dd>The minor allele frequency of the input variant (in this population).</dd>
 <dt>TAGVARIANT</dt><dd>The ID of the tagging variant.</dd>
 <dt>TAGCHR</dt><dd>The chromosome of the tagging variant.</dd>
 <dt>TAGSTART</dt><dd>The start position of the tagging variant.</dd>
 <dt>TAGEND</dt><dd>The end position of the tagging variant.</dd>
 <dt>POP</dt><dd>The population used to compute this r2 value.</dd>
 <dt>NGENOTYPES</dt><dd>The number of genotypes used to compute r2.
     This reflects taking the intersection of the called genotypes at the two variant sites in the specified population.</dd>
 <dt>NREFALLELES</dt><dd>The number of reference alleles observed in the input variant.</dd>
 <dt>TAGNREFALLELES</dt><dd>The number of reference alleles observed in the tagging variant.</dd>
 <dt>OVERLAPMAF</dt><dd>The minor allele frequency of the input variant measured only at overlapping samples with the tagging variant.
     This may be different than MAF, which considers all samples in this population for which there are input genotypes.</dd>
 <dt>TAGOVERLAPMAF</dt><dd>The minor allele frequency of the tagging variant measured only at overlapping samples with the input variant.</dd>
 <dt>R2</dt><dd>The squared correlation (r<sup>2</sup>) between the input and tagging variant.</dd>
 <dt>MAJOR</dt><dd>Integer specifying which allele in the input VCF file is higher frequency in this comparison.
       By comparing MAJOR and TAGMAJOR, you can tell whether the reference allele of the input variant corresponds to the reference or alternate allele of the tagging variant.
       Will be -1 for copy number variants.</dd>
 <dt>TAGMAJOR</dt><dd>Integer specifying which allele in the tagging VCF file is higher frequency in this comparison.<dd>
 </dl>
 
 <h3>Interpretation</h3>
 <p>
 It is important to evaluate the significance of a correlation based on the number of data points used to estimate the correlation.  If NGENOTYPES is low,
 many high r2 values would be expected by chance.
 </p><p>
 It is important to use reasonable window sizes, especially when evaluating the correlation of rare variants.  Singletons are by definition perfectly
 correlated with other singletons in the same sample.
 </p>
 
 <h3>Example</h3>
 <p>
 <pre>
 java -Xmx4g -cp SVToolkit.jar \
     org.broadinstitute.sv.main.SVAnnotator \
     -A TagVariants \
     -R human_g1k_v37.fasta \
     -vcf input.vcf \
     -tagFile tag_variants.vcf \
     -tagR2Threshold 0.5 \
     -populationMapFile 1000G_populations.map \
     -population CEU \
     -writeReport true \
     -writeSummary true \
     -reportDirectory reportdir
 </pre>

            <hr>
            <h3>TagVariantsAnnotator specific arguments</h3>
            <table class="table table-striped table-bordered table-condensed">
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Type</th>
                    <th>Default value</th>
                    <th>Summary</th>
                </tr>
                </thead>
                <tbody>
        <tr>
            <th colspan="4" id="row-divider">Optional</th>
        </tr>
            <tr>
                <td><a href="#--filterGenotypes">--filterGenotypes</a></td>
                <td>String</td>
                <td>NA</td>
                <td>True to ignore genotypes that have been filtered (default true).  Applies to both the input variants and tag variants.</td>
            </tr>
            <tr>
                <td><a href="#--filterVariants">--filterVariants</a></td>
                <td>String</td>
                <td>NA</td>
                <td>True to ignore variants that have been filtered (default true).  Applies to both the input variants and tag variants.</td>
            </tr>
            <tr>
                <td><a href="#--genotypeQualityThreshold">--genotypeQualityThreshold</a></td>
                <td>Double</td>
                <td>NA</td>
                <td>Ignore genotypes below this genotype quality GQ value (default no threshold).  Applies to both the input variants and tag variants.</td>
            </tr>
            <tr>
                <td><a href="#--population">--population</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Population(s) or .list file of populations to process</td>
            </tr>
            <tr>
                <td><a href="#--populationMapFile">--populationMapFile</a></td>
                <td>List[File]</td>
                <td>NA</td>
                <td>Map file (or files) containing sample to population assignments</td>
            </tr>
            <tr>
                <td><a href="#--tagIncudeOverlapping">--tagIncudeOverlapping</a></td>
                <td>String</td>
                <td>NA</td>
                <td>True to calculate r2 for overlapping variants (default false)</td>
            </tr>
            <tr>
                <td><a href="#--tagR2Threshold">--tagR2Threshold</a></td>
                <td>Double</td>
                <td>NA</td>
                <td>Minimum r2 value to report in report file (default no threshold)</td>
            </tr>
                </tbody>
            </table>



                        <hr>
                            <h2>Argument details</h2>
    <h3><a name="--filterGenotypes">--filterGenotypes
         / -filterGenotypes
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        True to ignore genotypes that have been filtered (default true).  Applies to both the input variants and tag variants.. 
            </p>
    <h3><a name="--filterVariants">--filterVariants
         / -filterVariants
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        True to ignore variants that have been filtered (default true).  Applies to both the input variants and tag variants.. 
            </p>
    <h3><a name="--genotypeQualityThreshold">--genotypeQualityThreshold
         / -genotypeQualityThreshold
    </a>
        (
        
        Double
        
        )
    </h3>
    <p class="args">
        Ignore genotypes below this genotype quality GQ value (default no threshold).  Applies to both the input variants and tag variants.. 
            </p>
    <h3><a name="--population">--population
         / -population
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Population(s) or .list file of populations to process. 
            </p>
    <h3><a name="--populationMapFile">--populationMapFile
         / -populationMapFile
    </a>
        (
        
        List[File]
        
        )
    </h3>
    <p class="args">
        Map file (or files) containing sample to population assignments. 
            </p>
    <h3><a name="--tagIncudeOverlapping">--tagIncudeOverlapping
         / -tagIncludeOverlapping
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        True to calculate r2 for overlapping variants (default false). 
            </p>
    <h3><a name="--tagR2Threshold">--tagR2Threshold
         / -tagR2Threshold
    </a>
        (
        
        Double
        
        )
    </h3>
    <p class="args">
        Minimum r2 value to report in report file (default no threshold). 
            </p>

        <hr>
        <p class="see-also">See also <a href="index.html">Documentation index</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip">GenomeSTRiP Site</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip/documentation">GenomeSTRiP Support</a></p>

        <p class="version">GenomeSTRiP version 1.04-r1441 built at 2014/07/14 14:26:06.</p>
        </div><!--/.span-->
        </div><!--/.row-->
        </body>
        <?php printFooter();
        ob_end_flush();
        ?>
