<!--
  ~ Copyright (c) 2013, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->

<!--
  ~ Copyright (c) 2012, The Broad Institute
  ~
  ~ Permission is hereby granted, free of charge, to any person
  ~ obtaining a copy of this software and associated documentation
  ~ files (the "Software"), to deal in the Software without
  ~ restriction, including without limitation the rights to use,
  ~ copy, modify, merge, publish, distribute, sublicense, and/or sell
  ~ copies of the Software, and to permit persons to whom the
  ~ Software is furnished to do so, subject to the following
  ~ conditions:
  ~
  ~ The above copyright notice and this permission notice shall be
  ~ included in all copies or substantial portions of the Software.
  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  ~ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  ~ OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  ~ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  ~ HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  ~ WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  ~ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  ~ OTHER DEALINGS IN THE SOFTWARE.
  -->



<?php $group = 'Variant Annotators'; ?>
        <?php ob_start(); ?>
        <!DOCTYPE html>
        <html lang='en'>
        <head>
            <meta charset='utf-8'>
            <title>AlleleFrequencyAnnotator documentation documentation</title>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <meta name='description' content=''>

            <link type="text/css" rel="stylesheet" href="bootstrap.min.css">
            <script src="jquery.min.js"></script>
            <script src="bootstrap.min.js"></script>
            <style>
                @media (min-width: 980px) {
                    body {
                        padding-top : 60px;
                    }
                }
            </style>

        </head>
        <body>
        <div class="navbar navbar-fixed-top">
            <div class="navbar-inner">
                <div class="container-fluid">
                    <a class="brand" href="index.html">GenomeSTRiP Documentation</a>
                </div>
            </div>
        </div>
        <div class="container-fluid">
            <?php
    // Output Buffering allows for the website to inject a different header
    ob_clean();
    include '../include/common.php';
    printHeader("AlleleFrequencyAnnotator documentation", "Guides");
    ?>

            <div class="row-fluid">

<section class="span4">
    <aside class="well">
        <h2>Categories</h2>
        <style>
            #sidenav .accordion-body a {
                color : gray;
            }

            .accordion-body li {
                list-style : none;
            }
        </style>
        <ul class="nav nav-pills nav-stacked" id="sidenav">
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#QueueScripts">Queue Scripts</a>

                    <div id="QueueScripts"
                    <?php echo ($group == 'Queue Scripts')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVDiscovery.html">SVDiscovery</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVGenotyper.html">SVGenotyper</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVPreprocess.html">SVPreprocess</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_qscript_SVQScript.html">SVQScript</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotation">Variant Annotation</a>

                    <div id="VariantAnnotation"
                    <?php echo ($group == 'Variant Annotation')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_SVVariantAnnotatorWalker.html">SVVariantAnnotatorWalker</a>
                                </li>
                    </ul>
                    </div>
                </li>
                <li><a data-toggle="collapse" data-parent="#sidenav" href="#VariantAnnotators">Variant Annotators</a>

                    <div id="VariantAnnotators"
                    <?php echo ($group == 'Variant Annotators')? 'class="accordion-body collapse in"'.chr(62) : 'class="accordion-body collapse"'.chr(62);?>
                    <ul>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_AlleleFrequencyAnnotator.html">AlleleFrequencyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ClusterSeparationAnnotator.html">ClusterSeparationAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_CopyNumberClassAnnotator.html">CopyNumberClassAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GCContentAnnotator.html">GCContentAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_GeneOverlapAnnotator.html">GeneOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_IntensityRankSumAnnotator.html">IntensityRankSumAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_NonVariantAnnotator.html">NonVariantAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_ReciprocalOverlapAnnotator.html">ReciprocalOverlapAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_RedundancyAnnotator.html">RedundancyAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_TagVariantsAnnotator.html">TagVariantsAnnotator</a>
                                </li>
                                <li>
                                    <a href="org_broadinstitute_sv_annotation_VariantsPerSampleAnnotator.html">VariantsPerSampleAnnotator</a>
                                </li>
                    </ul>
                    </div>
                </li>
        </ul>
    </aside>
    <?php getForumPosts( 'AlleleFrequencyAnnotator' ) ?>

</section>

<div class="span8">

    <h1>AlleleFrequencyAnnotator</h1>

    <p class="lead">Annotator that calculates summary statistics on population allele frequency and Hardy-Weinberg equilibrium.</p>
        <h3>Category
            <small> Variant Annotators</small>
        </h3>
    <br/>

    <h2>Introduction</h2>
    

 <p>
 The AlleleFrequency annotator is invoked through the <a href="org_broadinstitute_sv_main_SVAnnotator.html">SVAnnotator</a> framework, which defines arguments common to all annotators.
 </p><p>
 The AlleleFrequency annotator uses the genotype calls from the VCF file to compute allele frequency statistics.
 Only bi-allelic variants are currently supported and the GT/FT/GQ tags are used to determine the genotype values.
 </p><p>
 Multiple populations (potentially overlapping) may be specified and then the allele frequency statistics will be calculated
 on a per-population basis.  If no populations are defined, the samples in the VCF are treated as a single population.
 </p><p>
 The test for Hardy-Weinberg equilibrium is an exact method that calculates a two-sided Hardy-Weinberg p-value
 (Wigginton, Cutler and Abecasis, American Journal of Human Genetics, 2005 [PMID: 15789306], implementation copyright 2003 by Jan Wigginton and Goncalo Abecasis).
 </p>
 
 <h3>Input Formats</h3>
 <p>
 Population map files are tab delimited files with two columns.  The first column specifies the sample identifier and the second column
 specifies a population identifier.  A header line is optional, but if present the column names should be SAMPLE and POPULATION.
 </p><p>
 Multiple population map files may be provided and these may assign multiple populations to the same sample.  This allows multiple levels
 of population structure or overlapping populations to be described.
 </p>
 
 <h3>Output Formats</h3>
 <p>
 This annotator can produce the following outputs: report file.
 </p><p>
 This report file contains one line for each population for each variant.  The columns in the report file are:
 </p>
 <dl>
 <dt>ID</dt><dd>The ID of the variant.</dd>
 <dt>PLOIDY</dt><dd>The expected ploidy at this site.  Sites on sex chromosomes that are not uniformly diploid will be missing some statistics.</dd>
 <dt>POPULATION</dt><dd>The population identifier.</dd>
 <dt>AAF</dt><dd>The non-reference allele frequency.</dd>
 <dt>NALLELES</dt><dd>The total number of alleles (chromosomes) in the population.</dd>
 <dt>NNOCALLALLELES</dt><dd>The number of alleles (chromosomes) for which no genotype call is available.</dd>
 <dt>NREFALLELES</dt><dd>The number of alleles (chromosomes) that match the reference.</dd>
 <dt>NALTALLELES</dt><dd>The number of alleles (chromosomes) that are the non-reference allele.</dd>
 <dt>NGENOTYPES</dt><dd>The total number of individuals in the population.</dd>
 <dt>NNOCALL</dt><dd>The number of individuals for which no genotype call is available.</dd>
 <dt>NHOMREF</dt><dd>The number of individuals that are called homozygous for the reference allele.</dd>
 <dt>NHET</dt><dd>The number of individuals that are called heterozygous.</dd>
 <dt>NHOMALT</dt><dd>The number of individuals that are called homozygous for the alternate allele.</dd>
 <dt>HWEPVALUE</dt><dd>Hardy-Weinberg equilibrium p-value.</dd>
 </dl>
 
 <h3>Example</h3>
 <p>
 <pre>
 java -Xmx4g -cp SVToolkit.jar \
     org.broadinstitute.sv.main.SVAnnotator \
     -A AlleleFrequency \
     -R human_g1k_v37.fasta \
     -vcf input.vcf \
     -populationMapFile 1000G_populations.map \
     -writeReport true \
     -reportDirectory reportdir
 </pre>

            <hr>
            <h3>AlleleFrequencyAnnotator specific arguments</h3>
            <table class="table table-striped table-bordered table-condensed">
                <thead>
                <tr>
                    <th>Name</th>
                    <th>Type</th>
                    <th>Default value</th>
                    <th>Summary</th>
                </tr>
                </thead>
                <tbody>
        <tr>
            <th colspan="4" id="row-divider">Optional</th>
        </tr>
            <tr>
                <td><a href="#--filterGenotypes">--filterGenotypes</a></td>
                <td>String</td>
                <td>NA</td>
                <td>True to ignore genotypes that have been filtered (default true)</td>
            </tr>
            <tr>
                <td><a href="#--filterVariants">--filterVariants</a></td>
                <td>String</td>
                <td>NA</td>
                <td>True to ignore variants that have been filtered (default true)</td>
            </tr>
            <tr>
                <td><a href="#--genotypeQualityThreshold">--genotypeQualityThreshold</a></td>
                <td>Double</td>
                <td>NA</td>
                <td>Ignore genotypes below this genotype quality GQ value (default no threshold)</td>
            </tr>
            <tr>
                <td><a href="#--population">--population</a></td>
                <td>List[String]</td>
                <td>NA</td>
                <td>Population(s) or .list file of populations to process</td>
            </tr>
            <tr>
                <td><a href="#--populationMapFile">--populationMapFile</a></td>
                <td>List[File]</td>
                <td>NA</td>
                <td>Map file (or files) containing sample to population assignments</td>
            </tr>
                </tbody>
            </table>



                        <hr>
                            <h2>Argument details</h2>
    <h3><a name="--filterGenotypes">--filterGenotypes
         / -filterGenotypes
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        True to ignore genotypes that have been filtered (default true). 
            </p>
    <h3><a name="--filterVariants">--filterVariants
         / -filterVariants
    </a>
        (
        
        String
        
        )
    </h3>
    <p class="args">
        True to ignore variants that have been filtered (default true). 
            </p>
    <h3><a name="--genotypeQualityThreshold">--genotypeQualityThreshold
         / -genotypeQualityThreshold
    </a>
        (
        
        Double
        
        )
    </h3>
    <p class="args">
        Ignore genotypes below this genotype quality GQ value (default no threshold). 
            </p>
    <h3><a name="--population">--population
         / -population
    </a>
        (
        
        List[String]
        
        )
    </h3>
    <p class="args">
        Population(s) or .list file of populations to process. 
            </p>
    <h3><a name="--populationMapFile">--populationMapFile
         / -populationMapFile
    </a>
        (
        
        List[File]
        
        )
    </h3>
    <p class="args">
        Map file (or files) containing sample to population assignments. 
            </p>

        <hr>
        <p class="see-also">See also <a href="index.html">Documentation index</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip">GenomeSTRiP Site</a> | <a
                href="http://www.broadinstitute.org/software/genomestrip/documentation">GenomeSTRiP Support</a></p>

        <p class="version">GenomeSTRiP version 1.04-r1441 built at 2014/07/14 14:26:06.</p>
        </div><!--/.span-->
        </div><!--/.row-->
        </body>
        <?php printFooter();
        ob_end_flush();
        ?>
